File: ios/FaceDetectorFrameProcessorPlugin.swift
import Foundation
import VisionCamera
import MLKitFaceDetection
import MLKitVision

@objc(FaceDetectorFrameProcessorPlugin)
public class FaceDetectorFrameProcessorPlugin: FrameProcessorPlugin {
    
    private var faceDetector: FaceDetector?
    private var currentOptions: FaceDetectorOptions?
    
    // Auto mode options (stored separately as they don't affect detector)
    private var autoMode: Bool = false
    private var windowWidth: CGFloat = 1.0
    private var windowHeight: CGFloat = 1.0
    private var cameraFacing: String = "front"
    
    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
        super.init(proxy: proxy, options: options)
        print("[FaceDetectorPlugin] Initialized")
        updateDetectorOptions(options as? [String: Any] ?? [:])
    }
    
    private func updateDetectorOptions(_ options: [String: Any]) {
        let performanceMode = options["performanceMode"] as? String ?? "fast"
        let landmarkMode = options["landmarkMode"] as? String ?? "none"
        let contourMode = options["contourMode"] as? String ?? "none"
        let classificationMode = options["classificationMode"] as? String ?? "none"
        let minFaceSize = options["minFaceSize"] as? CGFloat ?? 0.15
        let trackingEnabled = options["trackingEnabled"] as? Bool ?? false
        
        // Update auto mode options
        autoMode = options["autoMode"] as? Bool ?? false
        windowWidth = options["windowWidth"] as? CGFloat ?? 1.0
        windowHeight = options["windowHeight"] as? CGFloat ?? 1.0
        cameraFacing = options["cameraFacing"] as? String ?? "front"
        
        let newOptions = FaceDetectorOptions()
        
        switch performanceMode {
        case "accurate":
            newOptions.performanceMode = .accurate
        default:
            newOptions.performanceMode = .fast
        }
        
        switch landmarkMode {
        case "all":
            newOptions.landmarkMode = .all
        default:
            newOptions.landmarkMode = .none
        }
        
        switch contourMode {
        case "all":
            newOptions.contourMode = .all
        default:
            newOptions.contourMode = .none
        }
        
        switch classificationMode {
        case "all":
            newOptions.classificationMode = .all
        default:
            newOptions.classificationMode = .none
        }
        
        newOptions.minFaceSize = minFaceSize
        newOptions.isTrackingEnabled = trackingEnabled
        
        faceDetector = FaceDetector.faceDetector(options: newOptions)
        currentOptions = newOptions
        print("[FaceDetectorPlugin] Detector updated with new options")
    }
    
    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
        if let args = arguments as? [String: Any], !args.isEmpty {
            updateDetectorOptions(args)
        }
        
        guard let detector = faceDetector else {
            print("[FaceDetectorPlugin] Detector not initialized")
            return []
        }
        
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(frame.buffer) else {
            print("[FaceDetectorPlugin] Could not get pixel buffer")
            return []
        }
        
        let image = VisionImage(buffer: frame.buffer)
        image.orientation = getImageOrientation(frame.orientation)
        
        let frameWidth = CVPixelBufferGetWidth(pixelBuffer)
        let frameHeight = CVPixelBufferGetHeight(pixelBuffer)
        
        var detectedFaces: [Face] = []
        let semaphore = DispatchSemaphore(value: 0)
        
        detector.process(image) { faces, error in
            if let error = error {
                print("[FaceDetectorPlugin] Detection error: \(error.localizedDescription)")
            } else if let faces = faces {
                detectedFaces = faces
            }
            semaphore.signal()
        }
        
        _ = semaphore.wait(timeout: .now() + 0.1)
        
        return detectedFaces.map { face in
            faceToDict(
                face, 
                frameWidth: frameWidth, 
                frameHeight: frameHeight,
                autoMode: autoMode,
                windowWidth: windowWidth,
                windowHeight: windowHeight,
                cameraFacing: cameraFacing
            )
        }
    }
    
    private func getImageOrientation(_ orientation: Orientation) -> UIImage.Orientation {
        switch orientation {
        case .portrait:
            return .right
        case .portraitUpsideDown:
            return .left
        case .landscapeLeft:
            return .up
        case .landscapeRight:
            return .down
        @unknown default:
            return .right
        }
    }
    
    private func faceToDict(
        _ face: Face, 
        frameWidth: Int, 
        frameHeight: Int,
        autoMode: Bool,
        windowWidth: CGFloat,
        windowHeight: CGFloat,
        cameraFacing: String
    ) -> [String: Any] {
        var result: [String: Any] = [:]
        
        // Calculate scale factors for autoMode
        let scaleX = autoMode ? windowWidth / CGFloat(frameWidth) : 1.0 / CGFloat(frameWidth)
        let scaleY = autoMode ? windowHeight / CGFloat(frameHeight) : 1.0 / CGFloat(frameHeight)
        let mirrorX = autoMode && cameraFacing == "front"
        
        // Bounding box
        let bounds = face.frame
        let x: CGFloat
        if mirrorX {
            x = windowWidth - ((bounds.origin.x + bounds.width) * scaleX)
        } else {
            x = bounds.origin.x * scaleX
        }
        
        result["bounds"] = [
            "x": x,
            "y": bounds.origin.y * scaleY,
            "width": bounds.width * scaleX,
            "height": bounds.height * scaleY
        ]
        
        // Angles
        result["rollAngle"] = face.headEulerAngleZ
        result["pitchAngle"] = face.headEulerAngleX
        result["yawAngle"] = face.headEulerAngleY
        
        // Classification probabilities
        if face.hasSmilingProbability {
            result["smilingProbability"] = face.smilingProbability
        }
        if face.hasLeftEyeOpenProbability {
            result["leftEyeOpenProbability"] = face.leftEyeOpenProbability
        }
        if face.hasRightEyeOpenProbability {
            result["rightEyeOpenProbability"] = face.rightEyeOpenProbability
        }
        
        // Tracking ID
        if face.hasTrackingID {
            result["trackingId"] = face.trackingID
        }
        
        // Landmarks
        var landmarks: [String: Any] = [:]
        
        if let leftEye = face.landmark(ofType: .leftEye) {
            landmarks["leftEye"] = pointToDict(leftEye.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let rightEye = face.landmark(ofType: .rightEye) {
            landmarks["rightEye"] = pointToDict(rightEye.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let noseBase = face.landmark(ofType: .noseBase) {
            landmarks["noseBase"] = pointToDict(noseBase.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let leftCheek = face.landmark(ofType: .leftCheek) {
            landmarks["leftCheek"] = pointToDict(leftCheek.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let rightCheek = face.landmark(ofType: .rightCheek) {
            landmarks["rightCheek"] = pointToDict(rightCheek.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let mouthLeft = face.landmark(ofType: .mouthLeft) {
            landmarks["mouthLeft"] = pointToDict(mouthLeft.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let mouthRight = face.landmark(ofType: .mouthRight) {
            landmarks["mouthRight"] = pointToDict(mouthRight.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let mouthBottom = face.landmark(ofType: .mouthBottom) {
            landmarks["mouthBottom"] = pointToDict(mouthBottom.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let leftEar = face.landmark(ofType: .leftEar) {
            landmarks["leftEar"] = pointToDict(leftEar.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        if let rightEar = face.landmark(ofType: .rightEar) {
            landmarks["rightEar"] = pointToDict(rightEar.position, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth)
        }
        
        if !landmarks.isEmpty {
            result["landmarks"] = landmarks
        }
        
        // All contours (like reference package)
        var contours: [String: Any] = [:]
        
        if let contour = face.contour(ofType: .face) {
            contours["face"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .leftEyebrowTop) {
            contours["leftEyebrowTop"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .leftEyebrowBottom) {
            contours["leftEyebrowBottom"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .rightEyebrowTop) {
            contours["rightEyebrowTop"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .rightEyebrowBottom) {
            contours["rightEyebrowBottom"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .leftEye) {
            contours["leftEye"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .rightEye) {
            contours["rightEye"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .upperLipTop) {
            contours["upperLipTop"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .upperLipBottom) {
            contours["upperLipBottom"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .lowerLipTop) {
            contours["lowerLipTop"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .lowerLipBottom) {
            contours["lowerLipBottom"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .noseBridge) {
            contours["noseBridge"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .noseBottom) {
            contours["noseBottom"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .leftCheek) {
            contours["leftCheek"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        if let contour = face.contour(ofType: .rightCheek) {
            contours["rightCheek"] = contour.points.map { pointToDict($0, scaleX: scaleX, scaleY: scaleY, mirrorX: mirrorX, windowWidth: windowWidth) }
        }
        
        if !contours.isEmpty {
            result["contours"] = contours
        }
        
        return result
    }
    
    private func pointToDict(
        _ point: VisionPoint, 
        scaleX: CGFloat, 
        scaleY: CGFloat,
        mirrorX: Bool,
        windowWidth: CGFloat
    ) -> [String: CGFloat] {
        let x: CGFloat
        if mirrorX {
            x = windowWidth - (point.x * scaleX)
        } else {
            x = point.x * scaleX
        }
        return [
            "x": x,
            "y": point.y * scaleY
        ]
    }
}

File: android/src/main/java/com/smartcamera/FaceDetectorFrameProcessorPlugin.kt
package com.smartcamera

import android.util.Log
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceContour
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetector
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy

// Data class to track current options for comparison
private data class DetectorConfig(
    val performanceMode: String,
    val landmarkMode: String,
    val contourMode: String,
    val classificationMode: String,
    val minFaceSize: Float,
    val trackingEnabled: Boolean,
    val autoMode: Boolean,
    val windowWidth: Float,
    val windowHeight: Float,
    val cameraFacing: String
)

class FaceDetectorFrameProcessorPlugin(proxy: VisionCameraProxy, options: Map<String, Any>?) : FrameProcessorPlugin() {
    
    companion object {
        private const val TAG = "FaceDetectorPlugin"
    }
    
    private var faceDetector: FaceDetector? = null
    private var currentConfig: DetectorConfig? = null
    
    init {
        Log.d(TAG, "FaceDetectorFrameProcessorPlugin initialized")
        updateDetectorOptions(options ?: emptyMap())
    }
    
    private fun updateDetectorOptions(options: Map<String, Any>) {
        val performanceMode = options["performanceMode"] as? String ?: "fast"
        val landmarkMode = options["landmarkMode"] as? String ?: "none"
        val contourMode = options["contourMode"] as? String ?: "none"
        val classificationMode = options["classificationMode"] as? String ?: "none"
        val minFaceSize = (options["minFaceSize"] as? Number)?.toFloat() ?: 0.15f
        val trackingEnabled = options["trackingEnabled"] as? Boolean ?: false
        val autoMode = options["autoMode"] as? Boolean ?: false
        val windowWidth = (options["windowWidth"] as? Number)?.toFloat() ?: 1.0f
        val windowHeight = (options["windowHeight"] as? Number)?.toFloat() ?: 1.0f
        val cameraFacing = options["cameraFacing"] as? String ?: "front"
        
        val newConfig = DetectorConfig(
            performanceMode = performanceMode,
            landmarkMode = landmarkMode,
            contourMode = contourMode,
            classificationMode = classificationMode,
            minFaceSize = minFaceSize,
            trackingEnabled = trackingEnabled,
            autoMode = autoMode,
            windowWidth = windowWidth,
            windowHeight = windowHeight,
            cameraFacing = cameraFacing
        )
        
        // Only recreate detector if core options changed
        if (currentConfig != null &&
            currentConfig!!.performanceMode == newConfig.performanceMode &&
            currentConfig!!.landmarkMode == newConfig.landmarkMode &&
            currentConfig!!.contourMode == newConfig.contourMode &&
            currentConfig!!.classificationMode == newConfig.classificationMode &&
            currentConfig!!.minFaceSize == newConfig.minFaceSize &&
            currentConfig!!.trackingEnabled == newConfig.trackingEnabled) {
            // Only update non-detector options
            currentConfig = newConfig
            return
        }
        
        val optionsBuilder = FaceDetectorOptions.Builder()
        
        when (performanceMode) {
            "accurate" -> optionsBuilder.setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
            else -> optionsBuilder.setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        }
        
        when (landmarkMode) {
            "all" -> optionsBuilder.setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
            else -> optionsBuilder.setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
        }
        
        when (contourMode) {
            "all" -> optionsBuilder.setContourMode(FaceDetectorOptions.CONTOUR_MODE_ALL)
            else -> optionsBuilder.setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
        }
        
        when (classificationMode) {
            "all" -> optionsBuilder.setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
            else -> optionsBuilder.setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
        }
        
        optionsBuilder.setMinFaceSize(minFaceSize)
        
        if (trackingEnabled) {
            optionsBuilder.enableTracking()
        }
        
        faceDetector?.close()
        faceDetector = FaceDetection.getClient(optionsBuilder.build())
        currentConfig = newConfig
        Log.d(TAG, "Face detector updated with new options")
    }
    
    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
        if (arguments != null && arguments.isNotEmpty()) {
            updateDetectorOptions(arguments)
        }
        
        val detector = faceDetector ?: run {
            Log.e(TAG, "Face detector not initialized")
            return emptyList<Map<String, Any>>()
        }
        
        val config = currentConfig ?: return emptyList<Map<String, Any>>()
        
        try {
            val mediaImage = frame.image
            val inputImage = InputImage.fromMediaImage(
                mediaImage,
                frame.orientation.toDegrees()
            )
            
            val faces = Tasks.await(detector.process(inputImage))
            
            return faces.map { face -> 
                faceToMap(
                    face, 
                    frame.width, 
                    frame.height,
                    config.autoMode,
                    config.windowWidth,
                    config.windowHeight,
                    config.cameraFacing
                ) 
            }
        } catch (e: Exception) {
            Log.e(TAG, "Face detection error: ${e.message}")
            return emptyList<Map<String, Any>>()
        }
    }
    
    private fun faceToMap(
        face: Face, 
        frameWidth: Int, 
        frameHeight: Int,
        autoMode: Boolean,
        windowWidth: Float,
        windowHeight: Float,
        cameraFacing: String
    ): Map<String, Any> {
        val result = mutableMapOf<String, Any>()
        
        // Calculate scale factors for autoMode
        val scaleX = if (autoMode) windowWidth / frameWidth else 1.0f / frameWidth
        val scaleY = if (autoMode) windowHeight / frameHeight else 1.0f / frameHeight
        val mirrorX = autoMode && cameraFacing == "front"
        
        // Bounding box
        val bounds = face.boundingBox
        val x = if (mirrorX) {
            windowWidth - (bounds.right * scaleX)
        } else {
            bounds.left * scaleX
        }
        
        result["bounds"] = mapOf(
            "x" to x.toDouble(),
            "y" to (bounds.top * scaleY).toDouble(),
            "width" to (bounds.width() * scaleX).toDouble(),
            "height" to (bounds.height() * scaleY).toDouble()
        )
        
        // Roll, pitch, yaw angles
        result["rollAngle"] = face.headEulerAngleZ.toDouble()
        result["pitchAngle"] = face.headEulerAngleX.toDouble()
        result["yawAngle"] = face.headEulerAngleY.toDouble()
        
        // Classification probabilities
        face.smilingProbability?.let { result["smilingProbability"] = it.toDouble() }
        face.leftEyeOpenProbability?.let { result["leftEyeOpenProbability"] = it.toDouble() }
        face.rightEyeOpenProbability?.let { result["rightEyeOpenProbability"] = it.toDouble() }
        
        // Tracking ID
        face.trackingId?.let { result["trackingId"] = it }
        
        // Landmarks
        val landmarks = mutableMapOf<String, Any>()
        
        face.getLandmark(FaceLandmark.LEFT_EYE)?.let {
            landmarks["leftEye"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.RIGHT_EYE)?.let {
            landmarks["rightEye"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.NOSE_BASE)?.let {
            landmarks["noseBase"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.LEFT_CHEEK)?.let {
            landmarks["leftCheek"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.RIGHT_CHEEK)?.let {
            landmarks["rightCheek"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.MOUTH_LEFT)?.let {
            landmarks["mouthLeft"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.MOUTH_RIGHT)?.let {
            landmarks["mouthRight"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.MOUTH_BOTTOM)?.let {
            landmarks["mouthBottom"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.LEFT_EAR)?.let {
            landmarks["leftEar"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        face.getLandmark(FaceLandmark.RIGHT_EAR)?.let {
            landmarks["rightEar"] = pointToMap(it.position.x, it.position.y, scaleX, scaleY, mirrorX, windowWidth)
        }
        
        if (landmarks.isNotEmpty()) {
            result["landmarks"] = landmarks
        }
        
        // All contours (like reference package)
        val contours = mutableMapOf<String, Any>()
        
        face.getContour(FaceContour.FACE)?.let { contour ->
            contours["face"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.LEFT_EYEBROW_TOP)?.let { contour ->
            contours["leftEyebrowTop"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.LEFT_EYEBROW_BOTTOM)?.let { contour ->
            contours["leftEyebrowBottom"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.RIGHT_EYEBROW_TOP)?.let { contour ->
            contours["rightEyebrowTop"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.RIGHT_EYEBROW_BOTTOM)?.let { contour ->
            contours["rightEyebrowBottom"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.LEFT_EYE)?.let { contour ->
            contours["leftEye"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.RIGHT_EYE)?.let { contour ->
            contours["rightEye"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.UPPER_LIP_TOP)?.let { contour ->
            contours["upperLipTop"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.UPPER_LIP_BOTTOM)?.let { contour ->
            contours["upperLipBottom"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.LOWER_LIP_TOP)?.let { contour ->
            contours["lowerLipTop"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.LOWER_LIP_BOTTOM)?.let { contour ->
            contours["lowerLipBottom"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.NOSE_BRIDGE)?.let { contour ->
            contours["noseBridge"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.NOSE_BOTTOM)?.let { contour ->
            contours["noseBottom"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.LEFT_CHEEK)?.let { contour ->
            contours["leftCheek"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        face.getContour(FaceContour.RIGHT_CHEEK)?.let { contour ->
            contours["rightCheek"] = contour.points.map { p -> pointToMap(p.x, p.y, scaleX, scaleY, mirrorX, windowWidth) }
        }
        
        if (contours.isNotEmpty()) {
            result["contours"] = contours
        }
        
        return result
    }
    
    private fun pointToMap(
        x: Float, 
        y: Float, 
        scaleX: Float, 
        scaleY: Float,
        mirrorX: Boolean,
        windowWidth: Float
    ): Map<String, Double> {
        val scaledX = if (mirrorX) {
            windowWidth - (x * scaleX)
        } else {
            x * scaleX
        }
        return mapOf(
            "x" to scaledX.toDouble(),
            "y" to (y * scaleY).toDouble()
        )
    }
    
    private fun com.mrousavy.camera.core.types.Orientation.toDegrees(): Int {
        return when (this) {
            com.mrousavy.camera.core.types.Orientation.PORTRAIT -> 0
            com.mrousavy.camera.core.types.Orientation.LANDSCAPE_LEFT -> 90
            com.mrousavy.camera.core.types.Orientation.PORTRAIT_UPSIDE_DOWN -> 180
            com.mrousavy.camera.core.types.Orientation.LANDSCAPE_RIGHT -> 270
        }
    }
    
    protected fun finalize() {
        faceDetector?.close()
    }
}

